library/core/src/unicode/printable/check.rs

```diff
@@ -0,0 +1,147 @@
+use crate::{iter, slice};
+
+/// A compact representation of Unicode singletons.
+///
+/// This is basically a `&[u16]`, but represented as `&[(u8, &[u8])]`,
+/// i.e. pairs of upper bytes and multiple corresponding lower bytes.
+///
+/// However, in order to reduce the pointer-sized overhead for each nested
+/// slice, it is compacted again into `&[(u8, u8)]` with the length of the
+/// lower bytes in the second byte, and a separate, contiguous `&[u8]` for
+/// storing the lower bytes.
+pub(super) struct Singletons {
+    upper: &'static [(u8, u8)],
+    lower: &'static [u8],
+}
+
+impl Singletons {
+    /// Creates a new `Singletons` instance from compacted upper and lower bytes.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the sum of all lengths (i.e. the second field of each pair) in `upper`
+    /// is not equal to the length of `lower`.
+    pub(super) const fn new(upper: &'static [(u8, u8)], lower: &'static [u8]) -> Self {
+        let mut lower_count_total = 0;
+        let mut i = 0;
+        while i < upper.len() {
+            lower_count_total += upper[i].1 as usize;
+            i += 1;
+        }
+        assert!(
+            lower_count_total == lower.len(),
+            "Sum of lengths in `upper` does not match `lower` length."
+        );
+
+        Self { upper, lower }
+    }
+
+    #[inline]
+    fn iter(&self) -> SingletonsIter {
+        SingletonsIter { iter: self.upper.iter().cloned(), lower: self.lower, lower_start: 0 }
+    }
+
+    pub(super) fn check(&self, x: u16) -> bool {
+        let [x_upper, x_lower] = x.to_be_bytes();
+        for (upper, lowers) in self.iter() {
+            if upper == x_upper {
+                for &lower in lowers {
+                    if lower == x_lower {
+                        return false;
+                    }
+                }
+            } else if x_upper < upper {
+                break;
+            }
+        }
+
+        true
+    }
+}
+
+struct SingletonsIter {
+    iter: iter::Cloned<slice::Iter<'static, (u8, u8)>>,
+    lower: &'static [u8],
+    lower_start: usize,
+}
+
+impl Iterator for SingletonsIter {
+    type Item = (u8, &'static [u8]);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let (upper, lower_count) = self.iter.next()?;
+
+        let lower_start = self.lower_start;
+        let lower_end = lower_start + lower_count as usize;
+        self.lower_start = lower_end;
+
+        // SAFETY: The invariant for `Singletons` guarantees that the sum of all lengths
+        // in `upper` must be equal to the lengths of `lower`, so `lower_end` is guaranteed
+        // to be in range.
+        let lowers = unsafe { self.lower.get_unchecked(lower_start..lower_end) };
```

Comment:
Do we have evidence that removing this unsafe has meaningful impact on performance or code size? I wouldn't expect the bounds checks to matter that much.

Reply:
@Mark-Simulacrum, the problem here isn't performance or code size, since on most targets, it is already optimized. The problem is that on some embedded targets, e.g. `thumbv7m-none-eabi`, the bounds check panic isn't optimized out in release mode, making it impossible to use `panic-never` or `no_panic` to ensure there are no reachable panics.

Reply:
Can you say more? Why do only some targets see the panics stay in?

In general, I don't think the standard library optimizes for "unreachable panics should not appear in the binary", and I'm reluctant to do so in this sort of code unless we gate that on e.g. a test-only assert and have exhaustive proof that it never triggers (e.g., by running it on every possible char).